File: 2026-01-January-GFG-POTD/January 07 - Count distinct elements in every window/Solution.cpp
class Solution{
public:
    vector<int> countDistinct(int arr[], int k, int n) {
        vector<int> ans;
        unordered_map<int,int> mp;

        // first window
        for(int i = 0; i < k; i++)
            mp[arr[i]]++;

        ans.push_back(mp.size());

        // slide window
        for(int i = k; i < n; i++){
            int outgoing = arr[i - k];
            mp[outgoing]--;
            if(mp[outgoing] == 0)
                mp.erase(outgoing);

            mp[arr[i]]++;
            ans.push_back(mp.size());
        }

        return ans;
    }
};

File: 2026-01-January-GFG-POTD/January 07 - Count distinct elements in every window/Solution.java
class Solution {
    ArrayList<Integer> countDistinct(int arr[], int k) {
        ArrayList<Integer> ans = new ArrayList<>();
        HashMap<Integer, Integer> map = new HashMap<>();
        
        // Build freq for first window
        for(int i = 0; i < k; i++){
            map.put(arr[i], map.getOrDefault(arr[i], 0) + 1);
        }
        ans.add(map.size());
        
        // Slide the window
        for(int i = k; i < arr.length; i++){
            int outgoing = arr[i - k];
            // decrease outgoing
            map.put(outgoing, map.get(outgoing) - 1);
            if(map.get(outgoing) == 0) map.remove(outgoing);
            
            // add new incoming
            map.put(arr[i], map.getOrDefault(arr[i], 0) + 1);
            
            ans.add(map.size());
        }
        
        return ans;
    }
}
